core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java

```diff
@@ -0,0 +1,358 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSplitHttpResponse.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscriber> downstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscriber.class, "downstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<DefaultSplitHttpResponse, HeadersFuture>
+            trailersFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultSplitHttpResponse.class, HeadersFuture.class, "trailersFuture");
+
+    private static final ResponseHeaders HEADERS_WITH_UNKNOWN_STATUS = ResponseHeaders.of(HttpStatus.UNKNOWN);
+    private static final HeadersFuture<HttpHeaders> EMPTY_TRAILERS;
+
+    static {
+        EMPTY_TRAILERS = new HeadersFuture<>();
+        EMPTY_TRAILERS.doComplete(HttpHeaders.of());
+    }
+
+    private final HeadersFuture<ResponseHeaders> headersFuture = new HeadersFuture<>();
+    private final BodySubscriber bodySubscriber = new BodySubscriber();
+    private final HttpResponse response;
+    private final EventExecutor executor;
+    private final boolean notifyCancellation;
+
+    @Nullable
+    private volatile HeadersFuture<HttpHeaders> trailersFuture;
+    private volatile boolean wroteAny;
+
+    public DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor,
+                                    SubscriptionOption... options) {
+        this.response = requireNonNull(response, "response");
+        this.executor = requireNonNull(executor, "executor");
+        notifyCancellation = containsNotifyCancellation(options);
+
+        response.subscribe(bodySubscriber, executor, options);
+    }
+
+    @Override
+    public final CompletableFuture<ResponseHeaders> headers() {
+        return headersFuture;
+    }
+
+    @Override
+    public final StreamMessage<HttpData> body() {
+        return this;
+    }
+
+    @Override
+    public final CompletableFuture<HttpHeaders> trailers() {
+        HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        if (trailersFuture != null) {
+            return trailersFuture;
+        }
+
+        trailersFuture = new HeadersFuture<>();
+        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
+            return trailersFuture;
+        } else {
+            return this.trailersFuture;
+        }
+    }
+
+    @Override
+    public boolean isOpen() {
+        return response.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return response.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor unused) {
+        requireNonNull(subscriber, "subscriber");
+        if (executor.inEventLoop()) {
```

Comment:
How about doing this after `downstreamUpdater.compareAndSet(this, null, downstream)`?